task/github-create-deployment/0.2/github-create-deployment.yaml

```diff
@@ -0,0 +1,187 @@
+    - name: STATUSES_URL
+      description: URL of the created deployment status.
+
+  steps:
+    - name: create-deployment
+      env:
+        - name: GITHUBTOKEN
+          valueFrom:
+            secretKeyRef:
+              name: $(params.GITHUB_TOKEN_SECRET_NAME)
+              key: $(params.GITHUB_TOKEN_SECRET_KEY)
+        - name: DEPLOYMENT_ID_PATH
+          value: $(results.ID.path)
+        - name: DEPLOYMENT_URL_PATH
+          value: $(results.URL.path)
+        - name: DEPLOYMENT_STATUSES_URL_PATH
+          value: $(results.STATUSES_URL.path)
+
+      image: registry.access.redhat.com/ubi8/ubi-minimal:8.2
+      script: |
+        #!/usr/libexec/platform-python
+        import http.client
+        import json
+        import os
+        import sys
+
+        deployment_url = "$(params.API_PATH_PREFIX)" + \
+            "/repos/$(params.REPO_FULL_NAME)/deployments"
+
+        if "$(params.REQUIRED_CONTEXTS)" == "*":
+            required_contexts = None
+        elif "$(params.REQUIRED_CONTEXTS)" == "":
+            required_contexts = []
+        else:
+            required_contexts = "$(params.REQUIRED_CONTEXTS)".split(",")
+
+        # payload should either be a JSON object or an empty string
+        if "$(params.PAYLOAD)":
+            payload = json.loads("$(params.PAYLOAD)")
+        else:
+            payload = ""
+
+        data = {
+            "ref": "$(params.REF)",
+            "task": "$(params.TASK)",
+            "auto_merge": json.loads("$(params.AUTO_MERGE)"),
+            "required_contexts": required_contexts,
+            "payload": payload,
+            "environment": "$(params.ENVIRONMENT)",
+            "description": "$(params.DESCRIPTION)",
+        }
+
+        print("Sending this data to GitHub: ")
+        print(data)
+
+        # This is for our fake github server
+        if "$(params.GITHUB_HOST_URL)".startswith("http://"):
+            conn = http.client.HTTPConnection("$(params.GITHUB_HOST_URL)".replace("http://", ""))
+        else:
+            conn = http.client.HTTPSConnection("$(params.GITHUB_HOST_URL)")
+
+        r = conn.request(
+            "POST",
+            deployment_url,
+            body=json.dumps(data),
+            headers={
+                "User-Agent": "TektonCD, the peaceful cat",
+                "Authorization": "Bearer " + os.environ["GITHUBTOKEN"],
+            })
+        resp = conn.getresponse()
+
+        # only look for 201 statuses to signify a deployment was successfully
+        # created, as a merged branch response returns a 202 status but does not
+        # actually create the deployment.
+        # https://developer.github.com/v3/repos/deployments/#merged-branch-response
+        if resp.status != 201:
+            print("Error: %d" % (resp.status))
+            print(resp.read())
+            sys.exit(1)
+        else:
+            body = json.loads(resp.read().decode())
+            open(os.environ.get('DEPLOYMENT_ID_PATH'), 'w').write(f'{body["id"]}')
```

Comment:
nit you don't need fstring here.

Reply:
Since type of `id` is `int`, passing raw value to `write` without string conversion will causes TypeError.

```
TypeError: write() argument must be str, not int
```

For `url`, `statuses_url`, it can be used as raw value because GitHub docs says it's type is `string`, but I'm not sure about using it (for safety). WDYT? 🤔

Reply:
ah yes you right!